src/reducers/index.js

```diff
@@ -105,102 +105,139 @@ function getInitialState() {
-      }
+        const scrollAnchor = getScrollAnchor(newState, newProps, oldProps);
 
-      newState = columnStateHelper.initialize(newState, newProps, oldProps);
+        // If anything has changed in state, update our rendered rows
+        if (!shallowEqual(state, newState) || scrollAnchor.changed) {
+          newState = computeRenderedRows(newState, scrollAnchor);
+        }
 
-      // if scroll values have changed, then we're scrolling!
-      if (newState.scrollX !== state.scrollX || newState.scrollY !== state.scrollY) {
-        newState.scrolling = newState.scrolling || true;
-      }
+        newState = columnStateHelper.initialize(newState, newProps, oldProps);
 
-      // TODO REDUX_MIGRATION solve w/ evil-diff
-      // TODO (jordan) check if relevant props unchanged and
-      // children column widths and flex widths are unchanged
-      // alternatively shallow diff and reconcile props
-      return newState;
-    }
-    case ActionTypes.SCROLL_END: {
-      const newState = Object.assign({}, state, {
-        scrolling: false,
-      });
-      const previousScrollAnchor = {
-        firstIndex: state.firstRowIndex,
-        firstOffset: state.firstRowOffset,
-        lastIndex: state.lastIndex,
-      };
-      return computeRenderedRows(newState, previousScrollAnchor);
-    }
-    case ActionTypes.SCROLL_TO_Y: {
-      let { scrollY } = action;
-      const newState = Object.assign({}, state, {
-        scrolling: true,
-      });
-      const scrollAnchor = scrollTo(newState, scrollY);
-      return computeRenderedRows(newState, scrollAnchor);
-    }
-    case ActionTypes.COLUMN_RESIZE: {
-      const { resizeData } = action;
-      return columnStateHelper.resizeColumn(state, resizeData);
-    }
-    case ActionTypes.COLUMN_REORDER_START: {
-      const { reorderData } = action;
-      return columnStateHelper.reorderColumn(state, reorderData);
-    }
-    case ActionTypes.COLUMN_REORDER_END: {
-      return Object.assign({}, state, {
-        isColumnReordering: false,
-        columnReorderingData: {}
-      });
-    }
-    case ActionTypes.COLUMN_REORDER_MOVE: {
-      const { deltaX } = action;
-      return columnStateHelper.reorderColumnMove(state, deltaX);
-    }
-    case ActionTypes.SCROLL_TO_X: {
-      const { scrollX } = action;
-      return Object.assign({}, state, {
-        scrolling: true,
-        scrollX,
-      });
-    }
-    default: {
-      return state;
+        // if scroll values have changed, then we're scrolling!
+        if (newState.scrollX !== state.scrollX || newState.scrollY !== state.scrollY) {
+          newState.scrolling = newState.scrolling || true;
+        }
+
+        // TODO REDUX_MIGRATION solve w/ evil-diff
+        // TODO (jordan) check if relevant props unchanged and
+        // children column widths and flex widths are unchanged
+        // alternatively shallow diff and reconcile props
+        return newState;
```

Comment:
Here and throughout: Since we're using Immer, can't we just mutate state directly, and avoid returning the new state? 
This should also avoid any cruft relating to copying fields.